tests/e2e/specs/subscriptions/purchase-sign-up-fee-subscription.spec.js

```diff
@@ -49,6 +56,39 @@ describeif( RUN_SUBSCRIPTIONS_TESTS )(
 			await fillCardDetails( page, card );
 			await shopper.placeOrder();
 			await expect( page ).toMatch( 'Order received' );
+
+			// Get the order ID so we can open it in the merchant view
+			const orderIdField = await page.$(
+				'.woocommerce-order-overview__order.order > strong'
+			);
+			orderId = await orderIdField.evaluate( ( el ) => el.innerText );
```

Comment:
If the next test depends on the order created here, they should both be combined into the same test case IMHO. Unless this is a common pattern in E2E tests.

Reply:
Ideally we want to keep the tests granular to help debug the test if it fails (see: [https://github.com/woocommerce/woocommerce/tree/trunk/tests/e2e#best-practices](https://github.com/woocommerce/woocommerce/tree/trunk/tests/e2e#best-practices)). So I wouldn't want to combine them together, and I think we get more value making sure we not only can place and order but we verified the charge came in as expected, too.

That said, we could have the second test create its own order to use so the test cases themselves are more atomic. What do you think?

Reply:
> Ideally we want to keep the tests granular to help debug the test if it fails (see: https://github.com/woocommerce/woocommerce/tree/trunk/tests/e2e#best-practices).

I can certainly see the value of having a "section" of a test fail. However, I've always been taught that test cases should be "atomic", there's even an option in most test runners to randomize the order in which tests are run so you can find and fix any hidden dependencies between tests. This test suite would fail in that scenario.

What you're doing here is basically a big test case that's separated into smaller "chunks" so it's easier to read. I don't have a lot of experience with our E2E tests so I don't know if that's an established pattern.

> That said, we could have the second test create its own order to use so the test cases themselves are more atomic. What do you think?

Ideally, yes. But if it's a significant amount of work, it's not worth it. I just wanted to bring to your attention that the test cases are not independent, but if you're OK with that let's merge this as-is.

Reply:
> But if it's a significant amount of work, it's not worth it. I just wanted to bring to your attention that the test cases are not independent, but if you're OK with that let's merge this as-is.

It wouldn't, and it's really easy to address. We would just need to add the order creation logic into the test case. The only downside is it will make the test suite itself run a bit longer as we need to create the order again in that test case.

Reply:
I ended up moving the order creation logic into the `beforeAll()` and dropping that step entirely. Instead we'll test for the charge and verify we have an active subscription after creating the order. Let me know if that works.

Reply:
It's strange having so much logic in the `beforeAll()` function. It's supposed to be for initialization only, but in this case it goes through a big chunk of the flow, including actually checking out (which is pretty much this test suite's purpose). I think I would extract that WCPay-specific logic to a separate function and call it at the start of every test case.

But I don't have a strong opinion now, sorry for making you rewrite it. I guess the "good practices" of unit tests can't apply completely to E2E. I'm fine with either this code or the way it was before. This one makes tests work independently but the previous code would run faster. It's your choice :)

Reply:
> But I don't have a strong opinion now, sorry for making you rewrite it

Don't be! This has been great feedback 👍 

> This one makes tests work independently but the previous code would run faster.

If it sounds good to you, I am thinking we can leave this as-is and come back and iterate on it to further optimize. For example, it would be great to be able to create subscription products via the API instead of the UI to help cut down on the overall test run times n the future.

Reply:
Yeah, sounds good 👍